File: python/compute_metrics.py
import numpy as np
from scipy.signal import hilbert, butter, filtfilt, welch

def loudness_bs1770(sig, fs):
    # Placeholder: integrate pyloudnorm in your pipeline
    return float(np.sqrt(np.mean(sig**2)))

def sharpness_proxy(sig, fs):
    X = np.fft.rfft(sig)
    f = np.fft.rfftfreq(len(sig), 1/fs)
    mag = np.abs(X)
    return float((f*mag).sum()/(mag.sum()+1e-12)/1000.0)

def roughness_proxy(sig, fs):
    env = np.abs(hilbert(sig))
    b,a = butter(4, [15,300], fs=fs, btype='band')
    mod = filtfilt(b,a,env)
    return float(np.sqrt(np.mean(mod**2)))

def fluctuation_proxy(sig, fs):
    env = np.abs(hilbert(sig))
    b,a = butter(4, 20, fs=fs, btype='low')
    slow = filtfilt(b,a,env)
    return float(np.var(slow))

def tonality_proxy(sig, fs):
    f,pxx = welch(sig, fs, nperseg=4096)
    base = np.convolve(pxx, np.ones(41)/41, mode='same')
    return float((pxx/base).max())

def annoyance(loud, sharp, rough, fluct):
    return float(loud*(1 + 0.25*sharp + 0.2*rough + 0.15*fluct))

def compute_all(sig, fs):
    L = loudness_bs1770(sig, fs)
    S = sharpness_proxy(sig, fs)
    R = roughness_proxy(sig, fs)
    F = fluctuation_proxy(sig, fs)
    T = tonality_proxy(sig, fs)
    A = annoyance(L,S,R,F)
    return dict(Loudness=L, Sharpness=S, Roughness=R, Fluctuation=F, Tonality=T, Annoyance=A)

File: matlab/compute_metrics.m
function results = compute_metrics(sig, fs)
% Compute psychoacoustic proxies (tutorial-style)
env = abs(hilbert(sig));
[bp_b,bp_a] = butter(4,[15 300]/(fs/2),'bandpass');
mod_env = filtfilt(bp_b,bp_a,env);
[lp_b,lp_a] = butter(4,20/(fs/2),'low');
slow_env = filtfilt(lp_b,lp_a,env);
[pxx,f] = pwelch(sig,4096,[],[],fs);
baseline = movmedian(pxx,41);

Loud = rms(sig);
Sharp = spectralCentroid(sig, fs)/1000;
Rough = rms(mod_env);
Fluct = var(slow_env);
Tonality = max(pxx./baseline);
Annoy = Loud*(1 + 0.25*Sharp + 0.2*Rough + 0.15*Fluct);

results = struct('Loudness',Loud,'Sharpness',Sharp,'Roughness',Rough, ...
    'Fluctuation',Fluct,'Tonality',Tonality,'Annoyance',Annoy);
end

function sc = spectralCentroid(x, fs)
X = abs(fft(x));
X = X(1:floor(numel(X)/2)+1);
f = (0:(numel(X)-1))*(fs/(2*(numel(X)-1)));
wsum = sum(X) + 1e-12;
sc = sum(f .* X) / wsum;
end
